src/lib/support/jsontlv/JsonToTlv.cpp

```diff
@@ -0,0 +1,418 @@
+        }
+        break;
+    }
+
+    case TLV::kTLVType_ByteString: {
+        VerifyOrReturnError(val.isString(), CHIP_ERROR_INVALID_ARGUMENT);
+        const std::string valAsString = val.asString();
+        size_t encodedLen             = valAsString.length();
+        VerifyOrReturnError(CanCastTo<uint16_t>(encodedLen), CHIP_ERROR_INVALID_ARGUMENT);
+
+        VerifyOrReturnError(isValidBase64String(valAsString), CHIP_ERROR_INVALID_ARGUMENT);
+
+        Platform::ScopedMemoryBuffer<uint8_t> byteString;
+        byteString.Alloc(BASE64_MAX_DECODED_LEN(static_cast<uint16_t>(encodedLen)));
+        VerifyOrReturnError(byteString.Get() != nullptr, CHIP_ERROR_NO_MEMORY);
+
+        auto decodedLen = Base64Decode(valAsString.c_str(), static_cast<uint16_t>(encodedLen), byteString.Get());
+        ReturnErrorOnFailure(writer.PutBytes(tag, byteString.Get(), decodedLen));
+        break;
+    }
+
+    case TLV::kTLVType_UTF8String: {
+        VerifyOrReturnError(val.isString(), CHIP_ERROR_INVALID_ARGUMENT);
+        ReturnErrorOnFailure(writer.PutString(tag, val.asCString()));
+        break;
+    }
+
+    case TLV::kTLVType_Null: {
+        VerifyOrReturnError(val.isNull(), CHIP_ERROR_INVALID_ARGUMENT);
+        ReturnErrorOnFailure(writer.PutNull(tag));
+        break;
+    }
+
+    case TLV::kTLVType_Structure: {
+        TLV::TLVType containerType;
+        VerifyOrReturnError(val.isObject(), CHIP_ERROR_INVALID_ARGUMENT);
+        ReturnErrorOnFailure(writer.StartContainer(tag, TLV::kTLVType_Structure, containerType));
+
+        std::vector<std::string> jsonNames = val.getMemberNames();
+        std::vector<ElementContext> nestedElementsCtx;
+
+        for (size_t i = 0; i < jsonNames.size(); i++)
+        {
+            ElementContext ctx;
+            ReturnErrorOnFailure(ParseJsonName(jsonNames[i], ctx));
+            nestedElementsCtx.push_back(ctx);
+        }
+
+        // Sort Json object elements by Tag number (low to high).
+        // Note that all sorted Context Tags will appear first floowed by all sorted Common Tags.
+        std::sort(nestedElementsCtx.begin(), nestedElementsCtx.end(), CompareByTag);
+
+        for (auto & ctx : nestedElementsCtx)
+        {
+            ReturnErrorOnFailure(EncodeTlvElement(val[ctx.jsonName], writer, ctx));
+        }
+
+        ReturnErrorOnFailure(writer.EndContainer(containerType));
+        break;
+    }
+
+    case TLV::kTLVType_Array: {
+        TLV::TLVType containerType;
+        VerifyOrReturnError(val.isArray(), CHIP_ERROR_INVALID_ARGUMENT);
+        ReturnErrorOnFailure(writer.StartContainer(tag, TLV::kTLVType_Array, containerType));
+
+        if (elementCtx.subType.tlvType == TLV::kTLVType_NotSpecified)
+        {
+            VerifyOrReturnError(val.size() == 0, CHIP_ERROR_INVALID_ARGUMENT);
+        }
+        else
+        {
+            ElementContext nestedElementCtx;
+            nestedElementCtx.type = elementCtx.subType;
```

Comment:
It would be clearer to explicitly set tag to anonymous on nestedElementCtx here.

Reply:
added